File: src/main/java/com/github/bjoernpetersen/jmusicbot/provider/DependencyMap.java
package com.github.bjoernpetersen.jmusicbot.provider;

import com.github.bjoernpetersen.jmusicbot.Plugin;
import java.util.Map;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public final class DependencyMap<T extends Plugin> {

  @Nonnull
  private final Map<Class<? extends T>, T> wrapped;

  public DependencyMap(@Nonnull Map<Class<? extends T>, T> wrapped) {
    this.wrapped = wrapped;
  }

  @Nullable
  public <Base extends T> Base get(Class<Base> baseClass) {
    return (Base) wrapped.get(baseClass);
  }
}

File: src/main/java/com/github/bjoernpetersen/jmusicbot/provider/DependencyMap.kt
package com.github.bjoernpetersen.jmusicbot.provider

import com.github.bjoernpetersen.jmusicbot.Plugin

interface DependencyMap<P : Plugin> {
  operator fun <Base : P> get(baseClass: Class<Base>): Base?
}

class DependencyMapWrapper<P : Plugin>(private val wrapped: Map<Class<out P>, P>) : DependencyMap<P> {
  override operator fun <Base : P> get(baseClass: Class<Base>): Base? {
    return baseClass.cast(wrapped[baseClass])
  }
}

class DependencyMapLookup<P : Plugin>(private val lookup: (Class<out P>) -> P?) : DependencyMap<P> {
  override operator fun <Base : P> get(baseClass: Class<Base>): Base? {
    return baseClass.cast(lookup(baseClass))
  }
}
